### ISSUE TITLE
Improve ShapeElementList performance.

### ISSUE THREAD

**Issue by pvcraven** [pvcraven/arcade#183]
## Enhancement request:

Just added shape_list_demo_1.py and shape_list_demo_2.py.

Demo 1 runs slow, and it should because no vertex buffers are used. Demo 2 runs 8 times faster, but it is still way too slow. Performance for non-changing shapes needs to be improved.

**Comment by zeziba** [comment]
Greetings again,

I was reading 
http://pyglet.readthedocs.io/en/pyglet-1.3-maintenance/programming_guide/graphics.html
and from what I am gathering is that if we stack the whole buffered object into one vertex list with indices this will do as you want. Currently each shape is creating it's own buffer and that means each buffer needs to have a call for drawing. It would take some work but if an object/shape is added to the buffered list it should return the vertices/object data for that object. Once that happens it can then stack the whole thing into a buffer/s and draw it as one.

Something like the following might work out.

```python

class ShapeElementList(Generic[T]):
    def __init__(self):
        ...
        self._buffer_array = None
        self._buffer_index = None

    def _create_buffer(self):
        from ctypes import pointer, sizeof

        self._buffer_array = gl.GLuint
        self._buffer_index = gl.GLuint

        # size of verticies
        _size = sum(len(l.vertex) for l in self.shape_list) * gl.GLfloat

        # vertex buffer
        glGenBuffers(1, pointer(self._buffer_array))
        gl.glBindBuffer(gl.GL_ARRAY_BUFFER, self._buffer_array)
        gl.glBufferData(gl.GL_ARRAY_BUFFER, _size, 0, gl.GL_STATIC_DRAW)

        _size_obj = 0
        for shape in self.shape_list:
            # to max size then create new buffer
            __size_obj = sum(len(shape.vertex))
            gl.glBufferSubData(gl.GL_ARRAY_BUFFER, _size_obj, _size_obj + __size_obj, shape)
            _size_obj = __size_obj

        # index buffer
        glGenBuffers(1, pointer(self._buffer_index))
        gl.glBindBuffer(gl.GL_ARRAY_BUFFER, self._buffer_index)
        gl.glBufferData(gl.GL_ARRAY_BUFFER, _size, 0, gl.GL_STATIC_DRAW)

        _size_obj = 0
        for shape in self.shape_list:
            # to max size then create new buffer
            __size_obj = sum(len(shape.vertex))
            gl.glBufferSubData(gl.GL_ELEMENT_ARRAY_BUFFER, _size_obj, _size_obj + __size_obj, shape)
            _size_obj = __size_obj

```

It would still be possible to translate and rotate scale and otherwise with a few modification. This would give the benefit of using one buffer object which should give the gains required.

Then drawing would be as simple as following each index.

The other solution is to build the vertex buffer as one when using the draw function then call the whole buffer with ```render``` or ```stripped_render``` or ```stripped_render_with_colors```. This would have the same effect but I need to look into it more to see what is simpler to implement and keep up to date.

I am going to try and modify the ```ShapeElementList.draw()``` to pass a array of all shapes to the ```render``` function to draw. This should make it markedly faster for many shapes.

Also, this should improve performance for static and dynamic shape lists completely across the board and any performance lag should be from the cpu/python bottleneck vs the draw functions. Afterwards it can be used on the regular draw commands potentially removing some of the delay for the commands.

Here is another sample of joining many shapes into one,
https://sites.google.com/site/swinesmallpygletexamples/vbo-triangle

**Comment by pvcraven** [comment]
Right. Colors would need to go into a list as well. I don't think it should be hard, just hasn't been done yet.

**Comment by zeziba** [comment]
I am playing with it right now, in pyglet there is a ```pyglet.graphics.Batch``` and it already does what is needed. 

Doing the following works with the supporting code

```
def draw_polygon_outline(point_list: PointList,
                         color: Color, border_width: float=1):
    if len(point_list) <= 0:
        return 
    gl.glEnable(gl.GL_BLEND)
    gl.glBlendFunc(gl.GL_SRC_ALPHA, gl.GL_ONE_MINUS_SRC_ALPHA)
    gl.glEnable(gl.GL_LINE_SMOOTH)
    gl.glHint(gl.GL_LINE_SMOOTH_HINT, gl.GL_NICEST)
    gl.glHint(gl.GL_POLYGON_SMOOTH_HINT, gl.GL_NICEST)

    # Set line width
    gl.glLineWidth(border_width)

    gl.glLoadIdentity()

    # Set color
    if len(color) == 4:
        gl.glColor4ub(color[0], color[1], color[2], color[3])
    elif len(color) == 3:
        gl.glColor4ub(color[0], color[1], color[2], 255)

    global GLOBAL_BATCH
    vlist = GLOBAL_BATCH

    l = tuple(round(p, 3) for point in point_list for p in point)
    # vlist.add(len(point_list), gl.GL_POLYGON, None, ("v2f", [p for point in point_list for p in point]))
    vlist.add(len(point_list), gl.GL_LINE_STRIP, None, ("v2f", [p for point in point_list for p in point]))
    vlist.add(2, gl.GL_LINE_STRIP, None, ("v2f", [*point_list[-1], *point_list[0]]))
```

The only problem I have working with now is trying to use a global batcher. Reading more into it it is possible to use the ```arcade.graphic.vertex_list``` with the ```batch``` in order to do this. There is streaming, dynamic and static attributes for the batcher allowing for optimizations of the objects. The only issue is that it is then required to keep a index location of the object in the batcher to update the vertices when needed.

Once I am done with my current little bit of code it should display what is needed to get this to work. 

I gathered this information from
http://pyglet.readthedocs.io/en/pyglet-1.3-maintenance/programming_guide/graphics.html

Also the batcher uses ```batch.draw()``` to draw the vertex set as well making it extremely simply in use after ll the hard work is ironed out.

**Comment by zeziba** [comment]
Here we go,
I have some code here to demonstrate the batcher from ```pyglet.graphic.Batcher```, it is drawing 10,000 lines all moving every frame.

```python
from random import random, randint, uniform

import arcade
import pyglet

SCREEN_WIDTH = 800
SCREEN_HEIGHT = 600


class Mover:
    def __init__(self, x, y, dx, dy, color, scale=1):
        self.x = x
        self.y = y
        self.dx = dx
        self.dy = dy
        self.color = color
        self.scale = scale
        self.large_inscribed_circle = 4
        self.small_inscribed_circle = 2
        self.angle = 0

    def update(self):
        if self.x < 0 or self.x > SCREEN_WIDTH:
            self.dx *= -1
        if self.y < 0 or self.y > SCREEN_HEIGHT:
            self.dy *= -1
        self.x += self.dx
        self.y += self.dy

    @property
    def position(self):
        return self.x - 5 * self.scale, self.y - 5 * self.scale, self.x + 5 * self.scale, self.y + 5 * self.scale


if __name__ == "__main__":
    num_objs = 10000
    
    @arcade.decorator.setup
    def setup(window):
        arcade.set_background_color(arcade.color.WHITE)
        window.batcher = pyglet.graphics.Batch()
        window.items = dict()
        for obj in [Mover(SCREEN_WIDTH * random(), SCREEN_HEIGHT * random(), random() * 5 + uniform(-1, 1),
                          random() * 5 + uniform(-1, 1),
                          (randint(0, 255), randint(0, 255), randint(0, 255))) for _ in range(0, num_objs)]:
            b = window.batcher.add(2, pyglet.gl.GL_LINES, None, ("v2f/stream", obj.position),
                                   ("c3b/static", obj.color * 2))
            window.items[hash(obj)] = {"object": obj, "vertices": b}


    @arcade.decorator.update
    def update(window, dt):
        for obj in window.items.keys():
            window.items[obj]["object"].update()
            window.items[obj]["vertices"].vertices = window.items[obj]["object"].position


    @arcade.decorator.draw
    def draw(window):
        window.batcher.draw()


    arcade.decorator.run(SCREEN_WIDTH, SCREEN_HEIGHT, title="Movers")

```

What is amazing is that it can go up to 20,000 with little problem but there is noticeable slow down at this point. At 200,000 it comes to a 2-3 fps on my laptop, but 10,000 objects being dynamically drawn is really good though.

The batcher returns an object that can be used to change the vertices of the object allowing for dynamic change of the object. This means that it should be easier to refactor the drawing commands to include this ability to speed everything up. It can be used on sprites and everything in-between allowing for some massive speed ups.

I am going to refactor my star demonstration to use this ability, an added bonus is that this should allow arcade to implement a particle system with little slowdown as well.

**Comment by zeziba** [comment]
Greetings once again,

I rewrote the example using the above methods and the results are what is expected and wanted. I was getting times of 0.002-0.003 per frame no matter the size of cubes I was making. Using a spacing of 3 and a size of 1 it was drawing a total of 52,272 rectangles onto screen.

Here is the new demo, I do not use any of the arcade methods as I believe they all need to be changed to incorporate the ```pyglet.graphic.Batcher``` to see these massive gains. Also the batcher takes sprites and primitives so all shapes should see and increase across the board if implemented.

The initial draw for the 50,000 squares takes roughly 1.4 secs though as a note.

```python
import pyglet
import arcade
import timeit


if __name__ == "__main__":
    SCREEN_WIDTH = 800
    SCREEN_HEIGHT = 600

    sqaure_width = 1
    square_height = 1
    sqaure_spacing = 3

    sqaure_color = arcade.color.DARK_GREEN
    background = arcade.color.DARK_BLUE_GRAY

    @arcade.decorator.setup
    def setup(window):
        arcade.set_background_color(background)
        window.batcher = pyglet.graphics.Batch()
        window.items = dict()
        window.draw_time = 0.0001
        sw = sqaure_width / 2.0
        sh = square_height / 2.0
        for x in range(sqaure_spacing, SCREEN_WIDTH, sqaure_spacing):
            for y in range(sqaure_spacing, SCREEN_HEIGHT, sqaure_spacing):
                x1 = x - sw
                x2 = x + sw
                y1 = y - sh
                y2 = y + sh
                window.batcher.add(4, pyglet.gl.GL_QUADS, None, ("v2f/static", [x1, y1, x2, y1, x2, y2, x1, y2]),
                                       ("c3b/static", sqaure_color * 4))


    @arcade.decorator.update
    def update(window, dt):
        pass


    @arcade.decorator.draw
    def draw(window):
        draw_time_start = timeit.default_timer()

        window.batcher.draw()
        output = "Drawing Time: {:.6f} seconds per frame".format(window.draw_time)

        label = pyglet.text.Label
        l = label(output,
                      font_name='Times New Roman',
                      font_size=18,
                      x=500, y=window.height - 40,
                      anchor_x='center', anchor_y='center')
        l.draw()

        window.draw_time = timeit.default_timer() - draw_time_start


    arcade.decorator.run(SCREEN_WIDTH, SCREEN_HEIGHT, title="Shapes of Many")

```

**Comment by zeziba** [comment]
To add more on this topic, I have abstracted the above to create squares.

Also as a note to the below code, all GL commands work before the object is added to the batch, this allows for some interesting recipes by the user. If where I am taking this is a good direction let me know and I will add sprites to this as well. Then after that particle effects can be added as it is only a small leap from here and would close out issue #196.

This also has the same draw time as the above of 0.002-0.004ish.

```python
import pyglet


class DrawObject:
    def __init__(self, points, gl_type, color, group, vertex_state="static", color_state="static"):
        self.points = points
        self.vertex_state = vertex_state
        self.color_state = color_state
        self.vertex_count = len(self.points)
        self.gl_type = gl_type
        self.group = group
        self.point_count = len(self.points)
        self.color = color if len(color) == self.point_count else [color] * self.point_count
        self.point_type = "v2f" if len(self.points[0]) == 2 else "v3f"
        self.color_type = "c3B" if len(self.color) == 3 else "c4B"


class Drawable:
    def __init__(self, draw_obj):
        self._draw_obj = draw_obj
        self._batch_obj = None

    def add(self, batch):
        if self._batch_obj is not None:
            return None
        drawable = self._draw_obj
        pnts = [p for point in drawable.points for p in point]
        cpnts = [p for point in drawable.color for p in point]
        self._batch_obj = batch.add(drawable.vertex_count, drawable.gl_type, drawable.group,
                                    ("v2f", pnts),
                                    ("c3B", cpnts))

    @property
    def vertices(self):
        return self._batch_obj.vertices

    @vertices.setter
    def vertices(self, points):
        self._batch_obj.vertices = points


class DrawableList:
    def __init__(self, *data):
        self.batch = pyglet.graphics.Batch()
        self.objects = {"objects": dict()}
        for obj in data:
            if type(obj) is DrawObject:
                self.add(Drawable(obj))
            else:
                print("Failed at {}, not the proper type".format(obj))

    def add(self, obj):
        h = hash(obj)
        self.objects["objects"][h] = obj
        return h

    def set_all(self):
        drawables = self.objects["objects"]
        for key in drawables.keys():
            drawables[key].add(self.batch)

    def draw(self):
        self.batch.draw()


def create_sqaure(x, y, width, height, color, group=None, dynamic=False, color_dynamic=False):
    x1 = x - width / 2.0
    x2 = x + width / 2.0
    y1 = y - height / 2.0
    y2 = y + height / 2.0
    return DrawObject([(x1, y1), (x2, y1), (x2, y2), (x1, y2)], pyglet.graphics.GL_QUADS, color, group,
                      "static" if dynamic is False else dynamic,
                      "static" if color_dynamic is False else color_dynamic)


if __name__ == "__main__":
    import arcade
    import timeit

    SCREEN_WIDTH = 800
    SCREEN_HEIGHT = 600

    sqaure_width = 1
    square_height = 1
    sqaure_spacing = 3

    sqaure_color = arcade.color.DARK_GREEN
    background = arcade.color.ASH_GREY


    @arcade.decorator.setup
    def setup(window):
        arcade.set_background_color(background)
        squares = []
        window.draw_time = 0.0001
        for x in range(sqaure_spacing, SCREEN_WIDTH, sqaure_spacing):
            for y in range(sqaure_spacing, SCREEN_HEIGHT, sqaure_spacing):
                squares.append(create_sqaure(x, y, sqaure_width, square_height, sqaure_color))
        window.batcher = DrawableList(*squares)
        window.batcher.set_all()


    @arcade.decorator.update
    def update(window, dt):
        pass


    @arcade.decorator.draw
    def draw(window):
        draw_time_start = timeit.default_timer()

        window.batcher.draw()
        output = "Drawing Time: {:.6f} seconds per frame".format(window.draw_time)

        label = pyglet.text.Label
        l = label(output,
                  font_name='Times New Roman',
                  font_size=18,
                  x=500, y=window.height - 40,
                  anchor_x='center', anchor_y='center')
        l.draw()

        window.draw_time = timeit.default_timer() - draw_time_start


    arcade.decorator.run(SCREEN_WIDTH, SCREEN_HEIGHT, title="Drawing New")
```

**Comment by pvcraven** [comment]
I've tried to mostly avoid using Pyglet except for OpenGL bindings, and some window management. I'd like to move away from Pyglet even for the window management part eventually. So using the 'batch' part of Pyglet isn't high on my priority list. I'd rather that batch code be run inside of Arcade. All the sprite stuff in Arcade does not use Pyglet's sprite or batching.

**Comment by zeziba** [comment]
OK, sounds good. I will look more into the batch though as it does what is done in https://github.com/pvcraven/arcade/blob/master/arcade/examples/shape_list_demo_3.py by sticking everything into one buffer and working from there. This should be able to be accomplished with the current shapelist as it only draws after each command set. Just throw it in a buffer and go from there.

I will check it out as it seems to be the next step for it.

**Comment by pvcraven** [comment]
Closed with 2.0 release.
